Feature/Transaction metadata

DeeJayElly opened the pull request:
Feature/Transaction metadata

Screenshots:

<img width="1454" alt="Screen Shot 2020-11-22 at 12 36 34 PM" src="https://user-images.githubusercontent.com/15862062/99902623-6c2a0180-2cbf-11eb-9b4a-d3037e785fd7.png">

<img width="1680" alt="Screen Shot 2020-11-09 at 2 13 00 PM" src="https://user-images.githubusercontent.com/15862062/98545866-68958400-2296-11eb-8e78-c172c63496e1.png">

<img width="1680" alt="Screen Shot 2020-11-09 at 2 13 08 PM" src="https://user-images.githubusercontent.com/15862062/98545860-6501fd00-2296-11eb-9f76-967a81c6f0d2.png">

<img width="1678" alt="Screen Shot 2020-12-10 at 11 29 44 AM" src="https://user-images.githubusercontent.com/15862062/101760470-3072a280-3adb-11eb-88cc-fdbabc28ea6a.png">

<img width="1672" alt="Screen Shot 2020-12-10 at 11 29 35 AM" src="https://user-images.githubusercontent.com/15862062/101760475-310b3900-3adb-11eb-861d-6d1fac6849ec.png">

--

## Test Cases

**Scenario 1: Transaction with metadata, show** 
Given I am on Cardano explorer app
And default language is ‘lang’
And I have a transaction with metadata
When I search for the transaction
Then I will see a dialog titled ’title_text’
And I will see the content ‘content_text’
When I select button 'show_text'
Then the dialog will be dismissed
And Metadata will be shown
And Metadata is formatted in readable manner

Example:
|lang|title_text|content_text|show_text|
|--|--|--|--|
|en|Warning: Unmoderated content|Transaction metadata is not moderated, and it may contain inappropriate content. The creator of the transaction provides this information and is not in control of this website's operator. Do you want to see the unmoderated content? |Show unmoderated content|
|jp|警告：非承認制コンテンツ|トランザクションメタデータは承認制ではなく、不適切な内容を含む場合があります。この情報はトランザクションの作成者が提供するもので、ウェブサイト管理者の管理下にはありません。管理されていないコンテンツを表示しますか。|管理されていないコンテンツを表示する|

**Scenario 2: Transaction with metadata, leave** 
Given I am on Cardano explorer app
And default language is ‘lang’
And I have a transaction with metadata
When I search for the transaction
Then I will see a button 'leave_text'
When I select the button
Then the dialog will be dismissed
And I am back on Cardano App search page

Example:
|lang|leave_text|
|--|--|
|en|Leave this page|
|jp|このページを離れる|

**Scenario 3: Transaction without metadata**
Given I am on Cardano explorer app
And default language is ‘lang’
And I have a transaction without metadata
When I search for the transaction
Then I will not see a dialog titled ‘title_text’

Example:
|lang|title_text|
|--|--|
|en|Warning: Unmoderated content|
|jp|警告：非承認制コンテンツ|

## Testing Summary

 For Hydra build #5131736:

|Test cases | Results|Note|
|-----------|----|----|
|Scenario 1 | Pass||
|Scenario 2| Pass||
|Scenario 3| Pass||
--

DeeJayElly commented:
@DominikGuzei @rhyslbw This PR is updated with new updates for Metadata feature. Pls review it.

DominikGuzei reviewed (changes requested):
@DeeJayElly the build fails due to a Typescript error: https://app.netlify.com/sites/cardano-explorer-mainnet/deploys/5fbb8804c4daca000821741f

DeeJayElly commented:
> @DeeJayElly the build fails due to a Typescript error: https://app.netlify.com/sites/cardano-explorer-mainnet/deploys/5fbb8804c4daca000821741f

Fixed @DominikGuzei @rhyslbw

DominikGuzei commented:
@rhyslbw please check again - I fixed the issues I have found and looks good now 👍

rhyslbw reviewed (changes requested):
@DeeJayElly 

I'm seeing an error: 

![Screenshot from 2020-11-30 20-40-00](https://user-images.githubusercontent.com/303881/100593273-531ef180-334c-11eb-89e5-88ebcedb3a2a.png)


Some transactions on the `mainnet` with metadata:

- `43c7a108ab69847ec2bdc3e283dca65a882271a41f01b19a008043494ff128be`
- `f910021138e553c65b96cf3e4647927fcd9f634e06544251f83cffb1891876e8`
- `2532ea3bcb91da8b40785aa6864fc65d022757219c7d406837e50daae475f163`

Also, please ensure the warning is only show if the transaction contains metadata

DeeJayElly commented:
> @DeeJayElly
> 
> I'm seeing an error:
> 
> ![Screenshot from 2020-11-30 20-40-00](https://user-images.githubusercontent.com/303881/100593273-531ef180-334c-11eb-89e5-88ebcedb3a2a.png)
> 
> Some transactions on the `mainnet` with metadata:
> 
> * `43c7a108ab69847ec2bdc3e283dca65a882271a41f01b19a008043494ff128be`
> * `f910021138e553c65b96cf3e4647927fcd9f634e06544251f83cffb1891876e8`
> * `2532ea3bcb91da8b40785aa6864fc65d022757219c7d406837e50daae475f163`
> 
> Also, please ensure the warning is only show if the transaction contains metadata

Fixed

DominikGuzei commented on source/features/i18n/translations/de.json:
```diff
@@ -161,13 +161,22 @@
+    "unmoderatedLeave": "!!!Leave this page",
+    "unmoderatedPoolTickerSymbols": "!!!Transaction metadata is not moderated and may contain inappropriate content.",
+    "unmoderatedTransactionData": "!!!Transaction metadata data is not moderated",
+    "unmoderatedWarning": "!!!Warning:"
```
What about the German translation? Do we still get this from someone?

DeeJayElly replied:
Added

DominikGuzei commented on source/features/address/ui/AddressSummary.tsx:
```diff
@@ -70,7 +76,9 @@ const AddressSummary = (props: AddressSummaryProps) => {
-                <div className={styles.infoValue}>{props.totalWithdrawn} ADA</div>
+                <div className={styles.infoValue}>
+                  {props.totalWithdrawn} ADA
+                </div>
```
Btw. why are half of the changes only formatting related? This is very confusing for reviews :(

DeeJayElly replied:
I am always using 'yarn prettier:format' before my commits/push . @DominikGuzei. I assume most of the old changes weren't formatted previously.

DominikGuzei commented on source/features/i18n/translations/ja.json:
```diff
@@ -161,13 +161,22 @@
+    "unmoderatedLeave": "!!!Leave this page",
+    "unmoderatedPoolTickerSymbols": "!!!Transaction metadata is not moderated and may contain inappropriate content.",
+    "unmoderatedTransactionData": "!!!Transaction metadata data is not moderated",
+    "unmoderatedWarning": "!!!Warning:"
```
What about Japanese?

DeeJayElly replied:
Added

DominikGuzei reviewed (changes requested):
@DeeJayElly I have left you some questions - but also can you briefly describe what you changed in your last commits? There have been a lot of formatting changes and it's hard to spot the "real" changes underneath

DeeJayElly commented:
> @DeeJayElly I have left you some questions - but also can you briefly describe what you changed in your last commits? There have been a lot of formatting changes and it's hard to spot the "real" changes underneath

@DominikGuzei I've fixed showing of unmoderated data dialog on transactions search page + i've added rendering of metadata in a slightly different way then rest of the rows on the transaction page.

https://github.com/input-output-hk/cardano-explorer-app/pull/372/commits/b738ccf9ed74a9d437da3de33503ffebb45087f4

So because metadata array can contain bunch of different types of data, objects, arrays, strings, etc... we are just formating them as a big JSON string:

`
{props.metadata.map(item => {
              return (
                <div>
                  {JSON.stringify(item.value)}
                </div>
              )
})}
`

rhyslbw commented on source/features/transactions/components/TransactionInfo.tsx:
```diff
@@ -256,6 +239,21 @@ const TransactionInfo = (props: ITransactionInfoProps) => {
+          </div>
+          <div className={styles.value}>
+            {props.metadata.map((item) => {
+              return <div>{JSON.stringify(item.value)}</div>;
```
```suggestion
              return <div>{JSON.stringify(item.value, undefined, 2)}</div>;
```

DeeJayElly replied:
Added

rhyslbw reviewed (changes requested):
1. The warning should **only** be shown if the transaction has metadata.
2. The formatting can be improved using my inline suggestion

DeeJayElly commented:
> 1. The warning should **only** be shown if the transaction has metadata.
> 2. The formatting can be improved using my inline suggestion

1. Fixed
2. Added

rhyslbw reviewed (changes requested):
Is there a styling override causing the `JSON.stringify` to not present the object with leading white space?

![Screenshot from 2020-12-02 00-15-26](https://user-images.githubusercontent.com/303881/100745737-d8310600-3433-11eb-87fa-34156f502ae9.png)

DeeJayElly commented:
> Is there a styling override causing the `JSON.stringify` to not present the object with leading white space?
> 
> ![Screenshot from 2020-12-02 00-15-26](https://user-images.githubusercontent.com/303881/100745737-d8310600-3433-11eb-87fa-34156f502ae9.png)

I inspected it, its just rendering like that all the time.

DeeJayElly commented:
> > Is there a styling override causing the `JSON.stringify` to not present the object with leading white space?
> > ![Screenshot from 2020-12-02 00-15-26](https://user-images.githubusercontent.com/303881/100745737-d8310600-3433-11eb-87fa-34156f502ae9.png)
> 



> Is there a styling override causing the `JSON.stringify` to not present the object with leading white space?
> 
> ![Screenshot from 2020-12-02 00-15-26](https://user-images.githubusercontent.com/303881/100745737-d8310600-3433-11eb-87fa-34156f502ae9.png)

Fixed @rhyslbw @DominikGuzei 

<img width="1675" alt="Screen Shot 2020-12-01 at 8 34 06 PM" src="https://user-images.githubusercontent.com/15862062/100787964-b6745680-3414-11eb-8573-19d6b21fe8b2.png">

DominikGuzei reviewed (changes requested):
@DeeJayElly @rhyslbw I talked to Darko and he said this:
"For stake pools, we need it once. But for metadata every time."

So I guess we need to change the logic for transactions to show the warning every time (only for tx that have the metadata) … maybe we should make this also less intrusive? Like, showing the warning only when the user actually wants to see the metadata?

rhyslbw reviewed (changes requested):
![Screenshot from 2020-12-07 21-55-16](https://user-images.githubusercontent.com/303881/101342645-f6d14a00-38d6-11eb-8271-f4c616e0602b.png)

DeeJayElly commented:
> @DeeJayElly @rhyslbw I talked to Darko and he said this:
> "For stake pools, we need it once. But for metadata every time."
> 
> So I guess we need to change the logic for transactions to show the warning every time (only for tx that have the metadata) … maybe we should make this also less intrusive? Like, showing the warning only when the user actually wants to see the metadata?

@DominikGuzei @rhyslbw I have implemented the requested change. Can you guys do another round of review.

rhyslbw reviewed (changes requested):
1. The warning must be shown every time, not once per session
2. Render the component in place of the metadata, so that the other details are shown

![Screenshot from 2020-12-09 12-26-05](https://user-images.githubusercontent.com/303881/101561882-43716e00-3a1a-11eb-9a54-f5875afb1a00.png)

DominikGuzei commented on source/features/search/ui/TransactionSearchResult.tsx:
```diff
@@ -37,14 +40,20 @@ export const TransactionSearchResult = () => {
-              showDetails
-              {...transactionSearchResult}
-            />
+            <>
```
i think you can remove the React Fragment wrapper here as the TX info is the only component to render

DeeJayElly replied:
Done

DominikGuzei commented on source/features/search/ui/TransactionSearchResult.tsx:
```diff
@@ -17,6 +17,9 @@ export const TransactionSearchResult = () => {
   const navigation = useNavigationFeature();
+  const [isShowingUnmoderatedData, setisShowingUnmoderatedData] = useState(
+    true
+  );
```
Is there a specific reason why the state for moderated data is kept in the search result and not directly in the `TransactionInfo` component?

DeeJayElly replied:
I're right, overcomplicating without a reason. I have simplified it now.

DeeJayElly commented:
> 1. The warning must be shown every time, not once per session
> 2. Render the component in place of the metadata, so that the other details are shown
> 
> ![Screenshot from 2020-12-09 12-26-05](https://user-images.githubusercontent.com/303881/101561882-43716e00-3a1a-11eb-9a54-f5875afb1a00.png)

Fixed

rhyslbw reviewed (dismissed):
Good result @DeeJayElly  :+1:

nikolaglumac commented:
@rhyslbw please take care of merging and deploying this change 🙏

mioriohk commented:
Hi @DeeJayElly. For Hydra build #5131736. Looks good to me. 👍 

|Test cases | Results|Note|
|-----------|----|----|
|Scenario 1 | Pass||
|Scenario 2| Pass||
|Scenario 3| Pass||
